### PR TITLE
DEPR: add stacklevel to FutureWarnings (GH9584)

### LINKED ISSUE #9584: FutureWarnings should specify a stacklevel
Running my test suite, I see:
```
/home/njs/.user-python2.7-64bit-3/local/lib/python2.7/site-packages/pandas/core/categorical.py:443: FutureWarning: Accessing 'levels' is deprecated, use 'categories'
  warn("Accessing 'levels' is deprecated, use 'categories'", FutureWarning)
/home/njs/.user-python2.7-64bit-3/local/lib/python2.7/site-packages/pandas/core/categorical.py:291: RuntimeWarning: Values and categories have different dtypes. Did you mean to use
'Categorical.from_codes(codes, categories)'?
  "'Categorical.from_codes(codes, categories)'?", RuntimeWarning)
/home/njs/.user-python2.7-64bit-3/local/lib/python2.7/site-packages/pandas/core/categorical.py:295: RuntimeWarning: None of the categories were found in values. Did you mean to use
'Categorical.from_codes(codes, categories)'?
  "'Categorical.from_codes(codes, categories)'?", RuntimeWarning)
./home/njs/.user-python2.7-64bit-3/local/lib/python2.7/site-packages/pandas/core/categorical.py:395: FutureWarning: 'labels' is deprecated. Use 'codes' instead
  warnings.warn("'labels' is deprecated. Use 'codes' instead", FutureWarning)
```

Okay, fair enough. But the problem is that all of the listed lines of code are inside pandas, so I have no idea which parts of my code are actually doing the wrong thing. `warnings.warn` provides a `stacklevel` argument, where the idea is that you should pass a non-zero value so that warnings are attributed to the code that actually needs to be changed. E.g., `stacklevel=1` means that the code which is doing something wrong is the code which called the function containing the call to `warnings.warn`.

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #9584

### CODE REVIEW

**Comment by jreback** [comment]
should revamp this....its pretty useful :)

**Comment by jorisvandenbossche** [comment]
yes, on my todo list! 
The problem I encountered is that for certain Warnings that are raised deeper into the pandas internals, the needed stacklevel is not always the same for different end user use cases. So it will not be possible to fix all of them. But for many (all warnings of deprecated functions or keywords, which are raised rather top-level), it is also simple to fix.

I now patched the `assert_produced_warning` to check for this, but should make it option so you can turn if off when it is not possible to get it right.

**Comment by jreback** [comment]
@jorisvandenbossche sounds good. Well at least the top-level ones would be a step up.

**Comment by jorisvandenbossche** [comment]
@jreback updated this. Added a check to `assert_produces_warning` that the stacklevel is set correctly, but for now only check it for FutureWarning/DeprecationWarning (all other warnings could be checked as well, but it was a bit too much work to get these all right for now).

File: pandas/computation/pytables.py [hunk 1]
```diff
@@ -535,7 +535,7 @@ def parse_back_compat(self, w, op=None, value=None):
                 w, op, value = w
             warnings.warn("passing a tuple into Expr is deprecated, "
                           "pass the where as a single string",
-                          DeprecationWarning)
+                          DeprecationWarning, stacklevel=10)
```
Comment (jreback) [inline pandas/computation/pytables.py]: whoosh deep nesting here!

**Comment by jreback** [comment]
lgtm.

what we actually need is a warning that then traverses the stack and finds the user level code as opposed to using an explict stack level, but I suspect thats a bit tricky.

**Comment by jorisvandenbossche** [comment]
> what we actually need is a warning that then traverses the stack and finds the user level code as opposed to using an explict stack level, but I suspect thats a bit tricky.

Hmm, that goes beyond my python skills, so for somebody else :-) (doubt if this is even possible)

Will merge tomorrow if no further comments

**Comment by jreback** [comment]
hahha

**Comment by jorisvandenbossche** [comment]
cc @njsmith since you reported this. This should solve almost all FutureWarnings we have in pandas.

**Comment by njsmith** [comment]
Doing that stack traversal thing is totally doable. Really it should be added as a feature upstream, like `warn(..., not_package="pandas.*")` to warn at the first level where the code isn't inside pandas (and there's some somewhat-relevant discussion in http://bugs.python.org/file40211). But in the mean time, woot, awesome.

**Comment by jorisvandenbossche** [comment]
Inferring the stacklevel is indeed not that difficult :-)

Something like this did it:
```
import warnings
import inspect

frame = inspect.currentframe()
stacklevel = 1

while '\\pandas\\' in frame.f_code.co_filename:
    frame = frame.f_back
    stacklevel += 1

warnings.warn('test!', stacklevel=stacklevel)
```

(the `'\\pandas\\'` is because I was testing on Windows)
But don't how robust something like this would be.

**Comment by jreback** [comment]
FYI I think u have to do a 

del frame at the end to break the gc cycles

**Comment by jreback** [comment]
@jorisvandenbossche let's merge this. I think I already changed a few as I thought you had merged this.

**Comment by jreback** [comment]
@jorisvandenbossche pls rebase and merge

**Pull request merged**